core/src/main/java/org/sonarsource/sonarlint/core/container/storage/SonarQubeActiveRulesProvider.java

```diff
@@ -21,57 +21,57 @@
 import org.sonarsource.sonarlint.core.container.analysis.SonarLintRules;
 import org.sonarsource.sonarlint.core.container.global.DefaultActiveRules;
 import org.sonarsource.sonarlint.core.container.standalone.rule.StandaloneActiveRuleAdapter;
 import org.sonarsource.sonarlint.core.container.standalone.rule.StandaloneRule;
-import org.sonarsource.sonarlint.core.serverapi.qualityprofile.QualityProfile;
 import org.sonarsource.sonarlint.core.serverapi.rules.ServerRules;
+import org.sonarsource.sonarlint.core.storage.ProjectStorage;
+import org.sonarsource.sonarlint.core.storage.RuleSet;
 
 public class SonarQubeActiveRulesProvider extends ProviderAdapter {
 
   private static final SonarLintLogger LOG = SonarLintLogger.get();
 
   private ActiveRules activeRules;
 
-  public ActiveRules provide(RulesStore rulesStore, List<QualityProfile> qProfiles, StorageReader storageReader, SonarLintRules rules,
-    ConnectedAnalysisConfiguration analysisConfiguration, Languages languages, ConnectedGlobalConfiguration globalConfiguration, ActiveRulesStore activeRulesStore) {
-    if (activeRules == null) {
-
-      Map<String, String> qProfilesByLanguage = loadQualityProfilesFromStorage(qProfiles, storageReader, analysisConfiguration);
+  public ActiveRules provide(SonarLintRules rules, RulesStore rulesStore, ConnectedAnalysisConfiguration analysisConfiguration,
+    Languages languages, ConnectedGlobalConfiguration globalConfiguration, ProjectStorage projectStorage) {
 
+    if (activeRules == null) {
       Collection<org.sonar.api.batch.rule.ActiveRule> activeRulesList = new ArrayList<>();
-      for (Map.Entry<String, String> entry : qProfilesByLanguage.entrySet()) {
-        String language = entry.getKey();
-        if (languages.get(language) == null) {
-          continue;
-        }
-
-        String qProfileKey = entry.getValue();
-        QualityProfile qProfile = qProfiles.stream().filter(qp -> qp.getKey().equals(qProfileKey)).findFirst().orElseThrow();
+      // could be empty before the first sync
+      var projectKey = analysisConfiguration.projectKey();
+      if (projectKey == null) {
+        // XXX should not happen, how to handle this case ?
+        LOG.debug("No project key provided, no rules will be used for analysis");
+        return new DefaultActiveRules(Collections.emptyList());
+      }
 
-        if (qProfile.getActiveRuleCount() == 0) {
-          LOG.debug("  * {}: '{}' (0 rules)", language, qProfile.getName());
+      var rulesByKey = rulesStore.getAll().stream().collect(Collectors.toMap(ServerRules.Rule::getRuleKey, Function.identity()));
+      for (Map.Entry<String, RuleSet> entry : projectStorage.getAnalyzerConfiguration(projectKey).getRuleSetByLanguageKey().entrySet()) {
+        String languageKey = entry.getKey();
+        var ruleSet = entry.getValue();
+        if (languages.get(languageKey) == null) {
+          // XXX is it possible ? active rules are already filtered by enabled languages
```

Comment:
It can maybe happen if a new language is enabled by the IDE, but the storage has not been refreshed yet.